lib/signs_ui_web/channels/signs_channel.ex

```diff
@@ -9,15 +9,53 @@ defmodule SignsUiWeb.SignsChannel do
 
   @spec handle_in(String.t(), %{Sign.id() => map()}, any()) :: {:noreply, Phoenix.Socket.t()}
   def handle_in("changeSigns", changes, socket) do
-    new_signs =
-      changes
-      |> Enum.map(fn {id, config} -> {id, Sign.from_config(id, config)} end)
-      |> Enum.into(%{})
+    if socket_authenticated?(socket) do
+      new_signs =
+        changes
+        |> Enum.map(fn {id, config} -> {id, Sign.from_config(id, config)} end)
+        |> Enum.into(%{})
 
-    {:ok, _new_state} = SignsUI.Signs.State.update_some(new_signs)
+      {:ok, _new_state} = SignsUI.Signs.State.update_some(new_signs)
 
-    Logger.info("Sign toggled: #{inspect(changes)}")
+      username = Guardian.Phoenix.Socket.current_resource(socket)
 
-    {:noreply, socket}
+      Logger.info("sign_changed: user=#{username} changes=#{inspect(changes)}")
+
+      {:noreply, socket}
+    else
+      {:noreply, maybe_send_auth_expired_message(socket)}
+    end
+  end
+
+  intercept(["sign_update"])
+
+  def handle_out("sign_update", msg, socket) do
+    if socket_authenticated?(socket) do
+      push(socket, "sign_update", msg)
+      {:noreply, socket}
+    else
+      {:noreply, maybe_send_auth_expired_message(socket)}
+    end
+  end
+
+  @spec socket_authenticated?(Phoenix.Socket.t()) :: boolean()
+  defp socket_authenticated?(socket) do
+    claims = Guardian.Phoenix.Socket.current_claims(socket)
+    token = Guardian.Phoenix.Socket.current_token(socket)
+
+    case SignsUiWeb.AuthManager.decode_and_verify(token, claims) do
+      {:ok, _claims} -> true
+      {:error, _error} -> false
+    end
+  end
+
+  @spec maybe_send_auth_expired_message(Phoenix.Socket.t()) :: Phoenix.Socket.t()
+  defp maybe_send_auth_expired_message(socket) do
+    if socket.assigns[:auth_expired_message_sent?] do
+      socket
+    else
+      :ok = push(socket, "auth_expired", %{})
```

Comment:
Should the socket get disconnected at this point as well?

Reply:
We experimented with having the `handle_in` / `handle_out` functions return `{:stop, ...}` in this case, but unfortunately that gets logged as a warning with no apparent way to reduce the severity. I suppose we could also close the socket on the front-end when it receives the `auth_expired` message, but it also calls `window.location.reload` in that handler so I'm not sure to what degree that would be an improvement.

Reply:
Yeah, as @lemald mentions, returning `{:stop, ...}` here was our first approach, but that terminated the channel and socket GenServers noisily in the logs. Perhaps we could go down the route of trapping exits and handling the termination messages? We didn't explore it too much since at this point they're not able to make any changes over the websocket and we wanted to send the frontend a message anyway, to trigger them to reload the page (and get redirected to Cognito).

Reply:
What was the full return you used? I'd figure `{:stop, :normal, _state}` would close the socket normally.

Reply:
Oh, you know what, I think we were using `{:stop, :authentication_expired}` or something to that effect. The Phoenix docs for the `handle_in` and `handle_out` callbacks don't mention anything about stopping normally, but you're right that the normal GenServer conventions probably apply. We'll try that.

Reply:
After a quick test, you're right: `{:stop, :normal, socket}` closes it normally without errors in the logs. We changed to that.